### PR TITLE
#0.20 skapa instansierbart random med nice funktioner

### PR DESCRIPTION
**Pull request by Manfred-Hastmark**
Created a instantiable random with same functionality as unity built in random

image

### CODE REVIEW

File: Assets/Scripts/Tools/ExtendedRandom.cs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+            else
+                return Value() * (a - b) + b;
+        }
+
+        /// <summary>
+        /// Returns random integer value
+        /// </summary>
+        public int Next()
+        {
+            return random.Next();
```
Comment (isakswe) [inline Assets/Scripts/Tools/ExtendedRandom.cs:125]: Random.next()'s max value is exclusive. Should be in the ranges of the different descriptions that depend on it. It also only returns positive unless a negative min is provided. Should be in description.
https://learn.microsoft.com/en-us/dotnet/api/system.random.next

File: Assets/Scripts/Tools/ExtendedRandom.cs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+            return new Quaternion(x, y, z, w).normalized;
+        }
+
+        /// <summary>
+        /// Returns a random float within [0, 1]
+        /// </summary>
+        public float Value()
+        {
+            return (float)random.NextDouble();
```
Comment (isakswe) [inline Assets/Scripts/Tools/ExtendedRandom.cs:98]: Random.NextDouble() has exclusive max, should be in description and ranges that depend on it. 
https://learn.microsoft.com/en-us/dotnet/api/system.random.nextdouble

File: Assets/Scripts/Tools/ExtendedRandom.cs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        {
+            float phi = Value(0, 2) * Mathf.PI; // Generates angle between [0, 2*Pi]
+            float theta = Value() * Mathf.PI; // Generates angle between [0, Pi]
+
+            // Convert from spherical coordinates to regular coordinates
+            return new Vector3(
+                x: Mathf.Sin(theta) * Mathf.Cos(phi),
+                y: Mathf.Sin(theta) * Mathf.Sin(phi),
+                z: Mathf.Cos(theta));
+        }
```
Comment (isakswe) [inline Assets/Scripts/Tools/ExtendedRandom.cs:78]: This makes points close to the center more likely. I guess we just ignore that? Other option would be randomising xyz within [-range, range] and checking if inside sphere.

File: Assets/Scripts/Tools/ExtendedRandom.cs (hunk 1 above)
Comment (Manfred-Hastmark) [reply Assets/Scripts/Tools/ExtendedRandom.cs:176]: Fixat

File: Assets/Scripts/Tools/ExtendedRandom.cs (hunk 2 above)
Comment (Manfred-Hastmark) [reply Assets/Scripts/Tools/ExtendedRandom.cs:149]: fixat

File: Assets/Scripts/Tools/ExtendedRandom.cs (hunk 3 above)
Comment (Manfred-Hastmark) [reply Assets/Scripts/Tools/ExtendedRandom.cs]: fixed

File: Assets/Scripts/Tools/ExtendedRandom.cs [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        /// </summary>
+        public Vector3 OnUnitSphere()
+        {
+            float phi = Value(0, 2) * Mathf.PI; // Generates angle between [0, 2*Pi)
+            float theta = Value() * Mathf.PI; // Generates angle between [0, Pi)
+
+            // Convert from spherical coordinates to regular coordinates
+            return new Vector3(
+                x: Mathf.Sin(theta) * Mathf.Cos(phi),
+                y: Mathf.Sin(theta) * Mathf.Sin(phi),
```
Comment (isakswe) [inline Assets/Scripts/Tools/ExtendedRandom.cs:111]: sin() for both theta and phi?

File: Assets/Scripts/Tools/ExtendedRandom.cs [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+        public Vector3 InsideUnitSphereUniform()
+        {
+            float phi = Mathf.Acos(Value(0, 2) - 1); // Generates angle between [0, 2*Pi)
+            float theta = Value(0, 2) * Mathf.PI; // Generates angle between [0, Pi)
+            float radius = Mathf.Pow(Value(), 1.0f / 3); // Generates radius between [0, 1]
+
+            // Convert from spherical coordinates to regular coordinates
+            return new Vector3(
+                x: radius * Mathf.Sin(phi) * Mathf.Cos(theta),
+                y: radius * Mathf.Sin(phi) * Mathf.Sin(theta),
```
Comment (isakswe) [inline Assets/Scripts/Tools/ExtendedRandom.cs:95]: sin() for both theta and phi? just making sure

File: Assets/Scripts/Tools/ExtendedRandom.cs [hunk 6]
```diff
@@ -63,12 +80,46 @@ public Vector3 InsideUnitSphere()
...
 
         /// <summary>
-        /// Returns a random point on the surface of a sphere with radius 1.0
+        /// <para>Returns a uniformly distributed random point inside or on a sphere with radius 1.0</para>
+        /// <i>Note: This function is slower than <see cref="InsideUnitSphere()">InsideUnitSphere()</see></i>
+        /// </summary>
+        public Vector3 InsideUnitSphereUniform()
+        {
+            float phi = Mathf.Acos(Value(0, 2) - 1); // Generates angle between [0, 2*Pi)
+            float theta = Value(0, 2) * Mathf.PI; // Generates angle between [0, Pi)
```
Comment (isakswe) [inline Assets/Scripts/Tools/ExtendedRandom.cs:89]: Descriptions switched. First is 0-Pi and second is 0-2Pi

File: Assets/Scripts/Tools/ExtendedRandom.cs [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+        }
+
+        /// <summary>
+        /// <para>Returns a uniformly distributed random point inside or on a sphere with radius 1.0</para>
+        /// <i>Note: This function is slower than <see cref="InsideUnitSphere()">InsideUnitSphere()</see></i>
+        /// </summary>
+        public Vector3 InsideUnitSphereUniform()
+        {
+            float phi = Mathf.Acos(Value(0, 2) - 1); // Generates angle between [0, 2*Pi)
+            float theta = Value(0, 2) * Mathf.PI; // Generates angle between [0, Pi)
```
Comment (isakswe) [inline Assets/Scripts/Tools/ExtendedRandom.cs:89]: Descriptions seem switched. First is 0-Pi and second is 0-2Pi

Comment (Manfred-Hastmark) [reply Assets/Scripts/Tools/ExtendedRandom.cs]: fixed

File: Assets/Scripts/Tools/ExtendedRandom.cs [hunk 8]
```diff
@@ -191,5 +179,13 @@ public int Next(int minValue, int maxValue)
         {
             return random.Next(minValue, maxValue);
         }
+
+        private Vector3 convertSphericalToCartesianCoords(float radius, float theta, float phi)
+        {
+            return new Vector3(
+                x: radius * Mathf.Sin(theta) * Mathf.Cos(phi),
+                y: radius * Mathf.Sin(theta) * Mathf.Sin(phi),
+                z: radius * Mathf.Cos(theta));
```
Comment (isakswe) [inline Assets/Scripts/Tools/ExtendedRandom.cs:188]: Seems like theta and phi are reversed. Other than that everything looks good.
New to spherical coordinates but looked at chapter "Converting from Spherical Coordinates" in https://math.libretexts.org/Bookshelves/Calculus/Book%3A_Calculus_(OpenStax)/12%3A_Vectors_in_Space/12.07%3A_Cylindrical_and_Spherical_Coordinates

Comment (Manfred-Hastmark) [reply Assets/Scripts/Tools/ExtendedRandom.cs:188]: It is correct as it is. The value which goes from 0->pi is, theta in this case is supposed to be where it is and the variable going from 0->2pi, phi in this case is supposed to be where it is. Though the normal notation is to have Theta going from 0->2pi, and phi going from 0->pi. I could change it if you really think it's important

Comment (isakswe) [reply Assets/Scripts/Tools/ExtendedRandom.cs:188]: All good. As long as you know what you're doing

**Pull request merged**